### PR TITLE
Add a higher order component to constrain Tab keyboard navigation.

### LINKED ISSUE #5242: Constrain tabbing within popovers and similar components
Follow up to #2306 see also #2323 that partially addressed keyboard navigation for this kind of components /cc @aduth 

There's one thing left to do though: treat this kind of components like sort of "modals" and constrain tabbing with the Tab key within the component.

As mentioned in #2306 
> Ideally, the Popover component should be placed in the source immediately after the controls that opens it. This way, it would be in the natural tab order and pressing tab would move focus to the Popover. 

Instead, now the Popover is mounted in the DOM very far from the control that opened it, typically close to the closing body tag. That's fine but it breaks the natural tab order, for example:
- using only the keyboard, use the Tab key to navigate to the Gutenberg top bar
- tab to the Inserter toggle (the "Add block" button) and press Enter
- the Inserter opens, focus is correctly moved to the Inserter
- tab through the Inserter 
- in a few tab key presses, you'll be outside of the Inserter and focus will be at the end of the document

This applies to various components, for example:
Inserter
Content structure
Visibility setting
Publish setting

This implementation makes using the interface with a keyboard very hard: great part of the interface gets basically "skipped", context is lost, and users are forced to start keyboard navigation from scratch or tab backwards.

In this kind of scenario, tabbing should be constrained within the component. 

Gutenberg has already the tools to get all the focusable / tabbable elements, so I'd suggest to consider to implement a new tool, probably a so called Higher Order Component, to constrain tabbing within components, where necessary.

This would be a first step to make popovers and the like usable with a keyboard. As a further improvement (to be addressed in a separate issue) we should evaluate if this kind of components need to be really treated as modals, that is: hiding all the other content from assistive technologies while they're open. For inspiration, see what [react-modal](https://github.com/reactjs/react-modal) does.

### PR DESCRIPTION
**Pull request by afercia**
## Description

This PR borrows code from #6261 special thanks to @Xyfi and aims to introduce a new Higher Order Component to constrain keyboard navigation with the Tab key.

Constraining tabbing is necessary for accessibility of various UI, for example modal dialogs, "popovers" and the like, where keyboard users must execute a specific task without navigating away and loosing context. For more details please refer to #5242.

TODO:

1
Consider to rename the component. The current name `withFocusContain` doesn't accurately describes what this component does. Actually, it's not about focus: it's about tabbing. Wondering if a different name would be beneficial, for example:
withTabbingContain?
withTabbingContained?
withTabbingConstrained?
Worth also considering the name should clarify this is something different from what other components do, for example`NavigableContainer`.

2
For now, and for testing purposes, this new HoC wraps the `Popover` component when it already has managed focus. Not sure this is a good assumption or if it's preferable to explicitly add this behavior via a specific prop. I guess this can always be done later in a new iteration.

3
There's a problem with the `focus.tabbables` implementation when the tabbables are exclusively radio buttons. This is clearly visible in the "Visibility setting" (see also screenshot below). Radio buttons are technically tabbables but it's actually possible to tab into just one of them. In this case, the implementation fails as `focus.tabbables` returns a collection of 3 elements but only one of them can be "tabbed" into:

screen shot 2018-05-28 at 17 17 14

Will open a separate issue.

For now, I'd recommend to test the behavior in some important pieces of the UI, most notably:
- the inserter
- the Content Structure popup
- the More ellipsis menu in the top bar 
- the link inline toolbar in an editable block
- Publish setting (the date picker)

To test:

- use only the keyboard and the Tab key to navigate 
- use Shift+Tab to navigate backwards 
- verify tabbing is always constrained within the component 
- verify all the functionalities of the wrapped component work as expected
- verify it's always possible to close the Popover pressing the Escape key

Fixes #5242

### CODE REVIEW

File: components/higher-order/with-focus-contain/index.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+					onKeyDown={ this.handleTabBehaviour }
+					ref={ this.focusContainRef }
+				>
+					<WrappedComponent { ...this.props } />
+				</div>
+			);
+			/* eslint-enable jsx-a11y/no-static-element-interactions */
+		}
+	};
+};
```
Comment (youknowriad) [inline components/higher-order/with-focus-contain/index.js]: I'm pretty sure a similar mechanism is also implemented elsewhere. Maybe NavigableContainer? Should we consolidate?

Comment (afercia) [reply components/higher-order/with-focus-contain/index.js]: > Worth also considering the name should clarify this is something different from what other components do, for example NavigableContainer.

🙂 Yes and no. The mechanism in NavigableContainer is used also for elements with tabindex="-1" like menu items.

**Review by aduth — APPROVED** [review]
I don't feel strongly about naming. `withConstrainedTabbing` would probably be my loosely-held preference.

Otherwise, this looks good 👍

**Comment by afercia** [comment]
Tested also in combination with the NUX tips PR see https://github.com/WordPress/gutenberg/pull/7503 seems to me it's working nicely. Coordinating with @Xyfi about https://github.com/WordPress/gutenberg/pull/6261 before merging.

**Pull request merged**